net: l2: ppp: ppp uart usage fixes

jhirsi opened the pull request:
This fixes 3 issues that came within PR #59124 for ppp uart usage.

Earlier start/stop of ppp was done at enable() but that was removed in PR #59124. 
Now putting enable/disable() back and putting start/stop there.

Additionally, there was a double ppp carrier ON when NET_EVENT_IF_DOWN. For that net_if_carrier_on/off is set in uart ppp.c driver. Also, worth to be mentioned that after PR #59124 there is no ppp carrier off when lcp is disconnected, for workaround that change, application that needs that information: should use ppp dead/running events instead.

bjarki-trackunit commented on subsys/net/l2/ppp/ppp_l2.c:
```diff
@@ -351,22 +350,30 @@ static void net_ppp_mgmt_evt_handler(struct net_mgmt_event_callback *cb, uint32_
-	}
-
 	if (mgmt_event == NET_EVENT_IF_UP) {
+#if defined(CONFIG_NET_PPP)
```
Starting and stopping the UART PPP wrapper (ppp.c) is the responsibility of the driver using it, see https://github.com/zephyrproject-rtos/zephyr/blob/5838b62f8914357f164467da7f0f447c20911fe4/drivers/modem/gsm_ppp.c#L629C28-L629C28 PPP L2 communicates in net_pkts, and should remain agnostic to the specific driver wrapping and transporting the net_pkts (in this case ppp.c CONFIG_NET_PPP).

rlubos replied:
But why should there be any extra action needed, to start/stop the ppp.c driver from the user? The ppp.c driver defines a network interface (https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/net/ppp.c#L1072), and when the application brings that interface UP, there should be no extra steps needed. That's why I've expressed my concern about removing `enable()` handler in the previous PR (https://github.com/zephyrproject-rtos/zephyr/pull/59124#discussion_r1226628581).

The standard path for all Zephyr network interfaces is:
* Application brings the interface up with `net_if_up()` 
* This triggers `l2_api->enable()` for that particular interface
* L2s `enable()` implementation should take care of any device-specific configuration needed to activate the interface. Just like for Ethernet, we call `start()/stop()` on the underlying Ethernet device, I can't really understand why shouldn't we do the same for PPP.

Right now I'm having the impression we're deviating from standard pattern of dealing with network interfaces, to address some specific, non standard behaviour of the `gsm_ppp.c` module.

jhirsi replied:
yes, saw that, it was done in a mentioned PR. Didn't fully understood that change. Was also wondering that net_if_up/down() calls in a driver? Not familiar with gsm_ppp, but maybe it is a driver for modem and more like application/boot up code for ppp net if for accomplishing modem driver tasks. Anyways, this would mean that if application wants to open ppp net if, it needs to start ppp 1st by using device api and then call net_if_up(). I do not feel comfortable with that.

Anyways, I'll let networking stack architects to judge this. My change for this shouldnt break gsm_ppp usage, eventhoug now different than in gsm_ppp usage.

bjarki-trackunit replied:
Remove the #if defined(CONFIG_NET_PPP) and the start/stop calls are fine.

We are currently using `net_if_carrier_on()/off()` to ensure the PPP L2 subsystem only establishes and drops LCP when the carrier is actually on/off, meaning the everything between the net_if send() API and the host is ready. For a modem, this means its registered to a network, and the device driver is ready to transport PPP frames to said modem.

This is how it works at this time for a modem:

1. Application powers on modem
2. Application brings up L2 network interface (like the one in ppp.c) using `net_if_up()`
3. While the modem is starting up, registering to a network, and having PPP set up, the PPP L2 subsystem does nothing, the carrier is not ready, so trying to start establishing LCP at this time will fail.
4. The modem registers to a network, and prepares PPP
5. The modem invokes `net_if_carrier_on()` to inform the L2 network interface that the carrier is in fact ready to receive net_pkts (wrapped in PPP frames but that is not relevant to the PPP L2 subsystem)
6. Since the carrier is now on, and the interface has been brought up (NET_IF_ADMIN_UP), the NET_IF_UP event is invoked by the net_if subsystem. This is what the PPP L2 subsystem reacts to https://github.com/zephyrproject-rtos/zephyr/blob/cbfa472fa7cfd7b98c8f853b5587337e0bcdc3f1/subsys/net/l2/ppp/ppp_l2.c#L346-L372
7. If the modem looses connection to the network, or is simply powered down, it will call `net_if_carrier_off()`. If this happens, or the application calls `net_if_down()`, the PPP L2 subsystem is informed, and LCP is dropped.

Its important to note that the carrier_on/carrier_off is related to the L2 network interface, not the higher layer LCP, IPCP etc. which relies on the L2 net_if. This seems to be where most of the confusion is coming from, we are not bringing up the L2 subsystem, we are bringing up the PPP data link layer.

bjarki-andreasen reviewed (changes requested):
In short, if you are using ppp.c, then you need to start it manually in your driver, before you call  net_if_carrier_on() on the network interface, (or net_if_up() as is done from within the gsm_ppp.c driver). If you need an event when LCP has established the data link, add the NET_EVENT_PPP_PHASE_RUNNING and NET_EVENT_PPP_PHASE_DEAD events where you have moved the CARRIER_ON/OFF events. 

This should solve your issues without adding complexity by mixing NET_PPP and NET_L2_PPP or changing the meaning of the CARRIER_ON/OFF events :)

jhirsi commented on subsys/net/l2/ppp/ppp_l2.c:
```diff
@@ -351,22 +350,30 @@ static void net_ppp_mgmt_evt_handler(struct net_mgmt_event_callback *cb, uint32_
 
 	ctx = net_if_l2_data(iface);
 
-	if (net_if_is_carrier_ok(iface)) {
```
Should be noted that with PPP, there are (unfortunately) still 2 separate carrier on/offs, net if (not really used with ppp afaik) and also PPP carrier on/off (have been and is used). Mentioned PR break/changed functionality of _PPP_ carrier on/off events. Changing/fixing that back more like it was, is one of purposes of this PR, not implementing net_if_carrier_on/off support.
Eventually, when implemented, net if carrier on/off could be like you said: indicating carrier on/off and hopefully we could get rid of ppp carrier on/off. The best would be that with ppp, NET_IF is notified to be usable only after IP is up (if IPCP used). But these are not the topic of this PR.

In my usage this highlighted code piece (condition for net_if_is_carrier_ok()) didn't work correctly as mixing 2 separate carrier on/offs here. Resulting odd PPP carrier events, for example NET_EVENT_IF_DOWN resulted PPP carrier ON.

bjarki-trackunit replied:
I agree, I left in the event in case it was being used somewhere downstream. This event should be called from a lower level which actually knows if the carrier is up/down, like a modem device driver which knows if the the data channel is ready and modem is registered to the network.

This does not mean that it should be placed in place of the proper events when LCP is running or dropped though right?

jhirsi replied:
As already said, this code here is mixing 2 separate issues net_if_carrier_on() is and has been different than PPP carrier ON.
Additionally, as you said, if understood correctly, PR #59124 creates undocumented dependency that needs to be met in order to work currently. And in your case, that "lower level" that sets net_if_carrier_on/off is in your own code, not in Zephyr code, right? At least I cannot find that in gsm_ppp.c

bjarki-trackunit replied:
The documentation describing how the L2/L1 interacts with the upper network stack is described here [[Network interface state management](https://docs.zephyrproject.org/latest/connectivity/networking/api/net_if.html)

The PR #59124 did not update the gsm_ppp.c driver to implement `net_if_carrier_on()/off()`. It was simply updated to not call the internal enable() function of the network interface directly. `enable()` is actually called by the net_if subsystem here.
https://github.com/zephyrproject-rtos/zephyr/blob/01ba923bb203c749f96bfa9d757f515d1fd91838/subsys/net/ip/net_if.c#L4356-L4391

The gsm_ppp.c driver never did and does not used the L2 network interface correctly. It always called net_if_up() internally as a result from a call to `gsm_ppp_start()` I did not want to break the driver for devs using it today, who expect `gsm_ppp_start()` and `gsm_ppp_stop()` to enable the modem (which is an incorrect design according to the documentation, and is irrelevant to the L2 PPP network interface and subsystem)

Quickly, I need to apologize that I may have stated `ppp_carrier_on()/off()` is used in the `gsm_ppp.c` driver, it is not and never has been. `ppp_carrier_on()/off()` is part of the network interface subsystem, and is in use across Zephyr.

I would recommend you closely study the changes in the PR #59124 to understand the broken state it was in before, and read the linked documentation before we continue the discussion.

jhirsi replied:
Yes, I've been really trying to find a tradeoff solution all the time. See updated PR and https://github.com/zephyrproject-rtos/zephyr/pull/61147#discussion_r1286966548

jhirsi committed:
net: l2: ppp: ppp uart usage fixed

This fixes 3 issues that came within PR #59124 for ppp uart usage.

Earlier start/stop of ppp was done at enable() but that
was removed in PR #59124. Now putting enable/disable() back and
putting start/stop there.
Additionally, there was a double ppp carrier ON when NET_EVENT_IF_DOWN.
For that net_if_carrier_on/off is set in uart ppp.c driver.
Also, maybe worth to be mentioned that after PR #59124 there is no
ppp carrier off when lcp is disconnected, for workaround that change,
application should use ppp dead/running events.

Signed-off-by: Jani Hirsimäki <jani.hirsimaki@nordicsemi.no>

rlubos reviewed (commented):
Thanks @jhirsi for the time you put into investigating this.

@bjarki-trackunit Please revisit, it seems now that the changes should be not be affecting your use case, while making `net/ppp.c` driver usable again.

bjarki-trackunit commented:
The fundamental issue with this PR is that it is trying to make the L2 PPP subsystem adhere to some downstream code, instead of updating the downstream code to adhere to the L2 PPP subsystem.

`ppp.c` is not part of the L2 PPP subsystem, it is simply one complex partial driver which depends on some other device driver to actually set up the carrier. Calling start() and stop() is not the responsibility of the L2 PPP subsystem, hence why it no longer does.

Additionally, the carrier is not LCP, when LCP is dead, it does not mean the carrier is lost. If some downstream code is misusing the event, that again, should be fixed there, not altered in the L2 PPP subsystem which accurately fires the event.

The changes in this PR simply do not adhere to the documentation and functionality of a L2 network interface and subsystem...

bjarki-trackunit commented:
Thank you for the reference.

The documentation is in this segment  [Network interface state management](https://docs.zephyrproject.org/latest/connectivity/networking/api/net_if.html)

To summarize enable() is called when a network interface is administratively brought up, not when it becomes operational.

```
struct net_l2 {
	...

	/**
	 * This function is used to enable/disable traffic over a network
	 * interface. The function returns <0 if error and >=0 if no error.
	 */
	int (*enable)(struct net_if *iface, bool state);

	...
};
```

If you tie the starting and stopping of the L2 driver to enable(), it will only be enabled if `net_if_up()` or `net_if_down()` is called. However, if the carrier is brought down, say a modem looses registration to the network, neither `net_if_up()` or `net_if_down()` is not called, `net_if_carrier_off()` is.

Once `net_if_carrier_off()` is called, LCP is terminated, and the L2 driver has to work to bring up the carrier again. Once the carrier is back, `net_if_carrier_on()` is called and the network interface becomes operational again.

Here is the issue: start() and stop() was never called, to reinitialize `ppp.c`, leaving it in an unknown state.

The only part of the system which knows when the carrier was acquired, lost, and reacquired is the L2 driver, which is why it is the only logical place to keep that logic.

I can not summarize the issue simpler than this. This PR provides a change which re-implements a fundamentally non-functional mechanism (while also adding `net_if_carrier_on()` and `net_if_carrier_off()` to `ppp.c` which results in the `net_if_up()` call mandating that the carrier is also on, which it can not know...)`

If we really want to keep the start and stop:
add start here
https://github.com/zephyrproject-rtos/zephyr/blob/f573fc9c40a2770843c5260af080eb7699f51058/subsys/net/l2/ppp/ppp_l2.c#L387
and stop here
https://github.com/zephyrproject-rtos/zephyr/blob/f573fc9c40a2770843c5260af080eb7699f51058/subsys/net/l2/ppp/ppp_l2.c#L392
you do not need to add anything in addition to this.

rlubos commented:
> The documentation is in this segment [Network interface state management](https://docs.zephyrproject.org/latest/connectivity/networking/api/net_if.html)

Good, I'm pretty well aware of that part of the documentation, as I wrote it.

> To summarize; enable() is called when a network interface is administratively brought up, not when it becomes operational.

And that's when it's supposed to be called, there's no confusion about that. Application brings the interface up (admin up), and the L2 starts the underlying device:
```c
struct ppp_api {
...
	/** Start the device */
	int (*start)(const struct device *dev);

	/** Stop the device */
	int (*stop)(const struct device *dev);
}
```

Don't you think that calling `start()/stop()` on the underlying network device after you enter/exit OPERATIONAL UP state on the interface, isn't logically sound? The interface should already be up and running, ready to transmit packets, yet you expect us to call the function which "Starts the device" then.

> If you tie the starting and stopping of the L2 driver to enable(), it will only be enabled if net_if_up() or net_if_down() is called. 

Is that wrong? If you bring the interface down, you stop the underlying device. And if you bring it up, you start it. What's wrong with that?

And it's not that you "tie" the API calls to `enable()` only - if your particular network device needs to be restarted in certain cases, you are still free to do so elsewhere. 

> Once net_if_carrier_off() is called, LCP is terminated, and the L2 driver has to work to bring up the carrier again. Once the carrier is back, net_if_carrier_on() is called and the network interface becomes operational again. At this point, LCP will reinitialize and **most likely fail**...

> Here is the issue: start() and stop() were never called, to reinitialize ppp.c, leaving it in an unknown state.

> The only part of the system which knows when the carrier was acquired, lost, and reacquired is the L2 driver, which is why it is the only logical place to keep that logic.

So if your particular modem driver needs to reboot the modem, do it, keep the logic there. But `ppp.c` is not a driver that has any modem awareness, nor a driver that needs reinitialization when link is lost (`start()` merely configures the UART). And most of all, the driver is **generic**, it allows to communicate between any two devices over PPP, not necessarily a modem.

> This PR provides a change which re-implements a fundamentally non-functional mechanism

Sorry, but even after reading all your comments here and in your previous PR that introduced the regression, I still don't understand what's non-fucntional about `enable()` in terms of PPP L2.

> (while also adding net_if_carrier_on() and net_if_carrier_off() to ppp.c which results in the net_if_up() call mandating that the carrier is also on, which it can not know...)`

I think we need to reconsider what a carrier means in terms of `ppp.c`. Yes, the driver is generic and it doesn't know that kind of device will be on the other end of the link. But the PHY (UART) is ready to transmit packets as soon as the UART driver is configured, and that's why carrier status should represent. So I see nothing wrong with the current change (in terms of `ppp.c` only).

jukkar reviewed (approved):
LGTM

bjarki-andreasen reviewed (changes requested):
One last retort, the UART being initialized is not the carrier being ready, what if the UART is not physically connected to anything?

bjarki-trackunit commented on subsys/net/l2/ppp/ppp_l2.c:
```diff
@@ -218,7 +217,31 @@ static enum net_l2_flags ppp_flags(struct net_if *iface)
+		net_if_get_device(iface)->api;
+	struct ppp_context *ctx = net_if_l2_data(iface);
+
+	if (ctx->is_enabled == state) {
```
orh nevermind, I'm wrong :D
